rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -0,0 +1,119 @@
+after the server has been provisioned which may be a slow process for
+organizations with large numbers of servers as it needs to be installed and then
+added to the teleport cluster
+
+With the changes described in this document, Teleport will be able to resolve
+the issues with adding AWS servers to Teleport clusters automatically.
+
+
+## Discovery
+
+A Teleport auth agent will need to be configured with an assume role policy for
+each of the accounts that the agent will discover services on as described by
+the EC2 join method docs[1]
+
+Discovery can use a matcher similar to the `db_service/aws` matcher:
+
+```yaml
+auth_service:
+  enabled: "yes"
+  aws:
+  - types: ["ec2"]
+    regions: ["us-west-1"]
+    tags:
+      "teleport": "yes"
+    accounts:
+    - aws_account: "222222222222"
+      aws_roles:
+      - "arn:aws:iam::222222222222:role/teleport-DescribeInstances-role"
+      - "arn:aws:iam::222222222222:role/teleport-Install-role"
+```
+
+The agent will use EC2's `DescribeInstances` API in order to list instances[1]. This
+will require the teleport auth agent to include `ec2:DescribeInstances` as part of
+it's IAM permissions
+
+As with database AWS discovery, new EC2 nodes will discovered periodically on a 5
+minute timer, and any new nodes will be added and existing instances updated/deleted.
+
+The EC2 join method already available in Teleport will be used to add nodes to the
+Teleport cluster, this sets the node name to the EC2 `<account-id>-<instance-id>`
+which can be reused here to avoid attempting to reinstall teleport on an instance
+where it is already present.
+Example:
+```json
+{
+  "kind": "node",
+  "version": "v2",
+  "metadata": {
+    "name": "AWS_INSTANCE_ID",
+    "labels": {
+      "env": "example"
+    },
+  },
+  "spec": {
+    "public_addr": "ec2-54-194-252-215.us-west-1.compute.amazonaws.com",
+    "hostname": "awsxyz"
+  }
+}
+```
+
+EC2 join tokens will have a `tags` field added so so only EC2 instances with matching
+`tags` can be added useing the token.
+
+## Agent installation
+
+In order to install the Teleport agent on EC2 instances, the
+`AWS-RunShellScript` document will be used to install and start teleport.
+
+In order to run `AWS-RunShellScript` the AWS user will need IAM permissions to
+run SSM commands [3] for example:
+
+```
+{
+    "Statement": [
+        {
+            "Action": "ssm:SendCommand",
+            "Effect": "Allow",
+            "Resource": [
+                "arn:aws:ssm:us-west-2:*:instance/*", # Allow running commands on all us-west-2 instances
+                "arn:aws:ssm:us-east-2:aws-account-ID:instance/i-02573cafcfEXAMPLE", # Allows running commands on specific instances in us-east
+                "arn:aws:ssm:us-east-2:aws-account-ID:document/AWS-RunShellScript" # Allows running shell scripts on resources
+            ]
+        }
+    ]
+}
+```
+
+The installation will run the host systems package manager to install Teleport
```

Comment:
Where would we store this script? What if users would like to tweak it? Is that going to be a part of the SSH service config?